api/src/main/java/org/eclipse/microprofile/rest/client/annotation/ClientHeaderParam.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (c) 2018 Contributors to the Eclipse Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.eclipse.microprofile.rest.client.annotation;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Repeatable;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * Used to specify an HTTP parameter that should be sent with the outbound request.
+ * When this annotation is placed at the interface level of a REST client interface, the specified header will be sent on each request for all
+ * methods in the interface.
+ * When this annotation is placed on a method, the header will be sent only for that method. If the same HTTP header is specified in an annotation
+ * for both the type and the method, only the header value specified in the annotation on the method will be sent.
```

Comment:
Does this need to cater to header params that are lists?

For instance, if the annotation is present on interface and method, if the type is a list then we add them both?

Reply:
I think that an annotation on the method should always override an annotation on the interface - this is consistent with other patterns (like Java EE's security annotations).

But, I think that you make a good point about headers that are lists.  We probably should make the value of the annotation a `String[]` rather than a `String`.

Reply:
Could we not just recognize it's a multi valued header and combine them?

Reply:
I think it would add complexity.  Potentially any header could have multiple values, so then we'd run into issues where it isn't clear whether a user is trying to override a header value in the method or whether they are trying to append another value.  We could add some sort of "appendable" attribute to the annotation to indicate that the value should be appended if multiple headers are already defined, but I think that makes it more complicated - and potentially unwieldy in a case like this:

```
@ClientHeaderParam(name="MyHeader", value="FromType")
public interface MyClient {

    @ClientHeaderParam(name="MyHeader", value="FromMethod")
    Response myMethod(@HeaderParam("MyHeader") passedInHeader);
}
...
myClient.myMethod("FromMethodArgument");
```

This could lead to sending:
`MyHeader: FromMethodArgument,FromType,FromMethod`
`MyHeader: FromType,FromMethod` (only using the ClientHeaderParams, ignoring the HeaderParam)
`MyHeader: FromMethodArgument` (only using the HeaderParam, ignoring the ClientHeaderParams)
or perhaps other combinations.

I was thinking that to make this as simple as possible, `@HeaderParam` takes priority over the `@ClientHeaderParam`, since that is passed-in by the user - part of the app.  Then the `@ClientHeaderParam` annotation on method would take priority over the annotation on the interface, since it is more specific.

So if a user wants multiple values for a particular header, then I think the annotation should support it with a `String[]` value attribute (and allow the default method to return a String array).  Alternatively, we could use multiple annotations at the same level - something like this:
```
@ClientHeaderParam(name="header1", value="value1")
@ClientHeaderParam(name="header1", value="value2")
@ClientHeaderParam(name="header1", value="value3")
Response myMethod();
```

but I prefer a String array attribute.

Reply:
Ok, I'd forgot about the override case.

So I'm good with making it a `String[]` and then following the original override means